### PR TITLE
Implement ldap configuration

### PR DESCRIPTION
**Pull request by yyorkchan**

### CODE REVIEW

File: pkg/util/validation/formats.go [hunk 1]
```diff
@@ -209,6 +216,137 @@ func (FormatHTTPOriginSpec) CheckFormat(value interface{}) error {
...
+		}
+	}
+	if err != nil {
+		return err
+	}
+	// check if the filter is correct
+	result := buf.String()
+	result = strings.TrimSpace(result)
+
+	_, err = ldap.CompileFilter(result)
```
[the remark this answers is not in the document]
Comment (cychiuae) [reply pkg/util/validation/formats.go:310]: The message is not readable and it expose too much implementation detail

File: pkg/lib/config/testdata/parse_secret_tests.yaml [hunk 2]
```diff
@@ -188,4 +188,24 @@ config:
...
+---
+name: ldap/missing-required-secret-key
+error: |-
+  invalid secrets:
+  /secrets/0/data: required
+    map[actual:<nil> expected:[name] missing:[name]]
+config:
+  secrets:
+    - key: ldap
+      data: {}
```
Comment (cychiuae) [inline pkg/lib/config/testdata/parse_secret_tests.yaml]: Try to be specific in the commit message.

> Update secret test yaml 

What are you updating? Everyone knows you are updating secret test yaml but for what.

File: pkg/lib/config/config.go [hunk 3]
```diff
@@ -89,6 +89,8 @@ type AppConfig struct {
 	Captcha       *CaptchaConfig       `json:"captcha,omitempty"`
 	BotProtection *BotProtectionConfig `json:"bot_protection,omitempty"`
 
+	LDAP *LDAPConfig `json:"ldap,omitempty"`
```
Comment (cychiuae) [inline pkg/lib/config/config.go:92]: I dont think this is a fix.

File: pkg/lib/config/secret.go [hunk 4]
```diff
@@ -215,6 +215,28 @@ func (c *SecretConfig) validateBotProtectionSecrets(ctx *validation.Context, bot
...
 }
 
+func (c *SecretConfig) validateLDAPServerUserSecrets(ctx *validation.Context, ldapServerConfig []LDAPServerConfig) {
+	c.validateRequire(ctx, LDAPServerUserCredentialsKey, "LDAP server user credentials")
+	_, data, _ := c.LookupDataWithIndex(LDAPServerUserCredentialsKey)
+	ldapServerUserCredentials, ok := data.(*LDAPServerUserCredentials)
+	if ok {
+		for _, server := range ldapServerConfig {
+			matched := false
+			for _, user := range ldapServerUserCredentials.Items {
```
Comment (cychiuae) [inline pkg/lib/config/secret.go:225]: It is not a `user`. Please rename it.

File: pkg/lib/config/secret_ldap.go [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+type LDAPServerUserCredentials struct {
+	Items []LDAPServerUserCredentialsItem `json:"items,omitempty"`
+}
+
+var _ = SecretConfigSchema.Add("LDAPServerUserCredentialsItem", `
+{
+	"type": "object",
+	"additionalProperties": false,
+	"properties": {
+		"name": { "type": "string" },
```
Comment (cychiuae) [inline pkg/lib/config/secret_ldap.go:22]: Do we need to add `minLength: 1` checking to name?

File: pkg/lib/config/testdata/parse_secret_tests.yaml [hunk 6]
```diff
@@ -188,4 +188,66 @@ config:
...
+  invalid secrets:
+  /secrets/0/data/items/0: dependentRequired
+    map[Required:map[dn:map[actual:[dn name] expected:[password] missing:[password]]]]
+config:
+  secrets:
+    - key: ldap
+      data:
+        items:
+          - name: test
+            dn: dc=test
```
Comment (cychiuae) [inline pkg/lib/config/testdata/parse_secret_tests.yaml:253]: End this file will a newline

File: pkg/lib/config/testdata/secret_config_validate_tests.yaml [hunk 7]
```diff
@@ -374,4 +374,99 @@ secret_config:
...
+          search_filter_template: (mail={{.Username}})
+          user_id_attribute_oid: "0.9.2342.19200300.100.1.1"
+secret_config:
+  secrets:
+    - key: ldap
+      data:
+        items:
+        - name: default
+          dn: "cn=admin,dc=localhost"
+          password: "password"
```
Comment (cychiuae) [inline pkg/lib/config/testdata/secret_config_validate_tests.yaml:472]: End this file will a newline

File: pkg/lib/config/ldap.go [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+{
+	"type": "object",
+	"additionalProperties": false,
+	"properties": {
+		"servers": {
+			"type": "array",
+			"items": {
+				"$ref": "#/$defs/LDAPServerConfig"
+			},
+			"minItems": 1
```
Comment (louischan-oursky) [inline pkg/lib/config/ldap.go:13]: Remove this. If you reference OAuth providers, there is no `minItems` there.

File: pkg/lib/config/ldap.go [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+
+type LDAPServerConfig struct {
+	Name                             string   `json:"name,omitempty"`
+	URL                              *url.URL `json:"url,omitempty"`
+	BaseDN                           string   `json:"base_dn,omitempty"`
+	SearchFilterTemplate             string   `json:"search_filter_template,omitempty"`
+	UserUniqueIdentifierAttributeOID string   `json:"user_id_attribute_oid,omitempty"`
+}
+
+var _ = SecretConfigSchema.Add("LDAPServerUserCredentials", `
```
Comment (louischan-oursky) [inline pkg/lib/config/ldap.go]: Move secret related things in `secret_ldap.go`. Take `bot_protection.go` and `secret_bot_protection.go` as references.

File: pkg/lib/config/ldap.go [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+}
+
+var _ = SecretConfigSchema.Add("LDAPServerUserCredentialsItem", `
+{
+	"type": "object",
+	"additionalProperties": false,
+	"properties": {
+		"name": { "type": "string" },
+		"dn": { "type": "string", "format": "ldap_dn" },
+		"password": { "type": "string" }
```
Comment (louischan-oursky) [inline pkg/lib/config/ldap.go]: `dn` and `password` should have `minLength: 1`

File: pkg/lib/config/ldap.go [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+	SearchFilterTemplate             string   `json:"search_filter_template,omitempty"`
+	UserUniqueIdentifierAttributeOID string   `json:"user_id_attribute_oid,omitempty"`
+}
+
+var _ = SecretConfigSchema.Add("LDAPServerUserCredentials", `
+{
+	"type": "object",
+	"additionalProperties": false,
+	"properties": {
+		"items": { "type": "array", "items": { "$ref": "#/$defs/LDAPServerUserCredentialsItem" } }
```
Comment (louischan-oursky) [inline pkg/lib/config/ldap.go]: Add programmatic validation to ensure if a server is defined, then it has a corresponding item in the secrets. You can also reference bot protection on how to do this.

File: pkg/util/validation/formats.go [hunk 12]
```diff
@@ -209,6 +216,135 @@ func (FormatHTTPOriginSpec) CheckFormat(value interface{}) error {
...
+	if err != nil {
+		return err
+	}
+	var buf bytes.Buffer
+	// make a string array of phone, username, email
+	testcases := []string{"+85298765432", "username", "test@test.com"}
+	for _, testcase := range testcases {
+		err = tmpl.Execute(&buf, map[string]string{"Username": testcase})
+		if err == nil && buf.String() != "" {
+			break
```
Comment (louischan-oursky) [inline pkg/util/validation/formats.go:301]: The search filter template MUST pass all three test cases in order to be considered as valid. So the loop should look like

```golang
testcases := []string{"+85298765432", "username", "test@test.com"}
for _, testcase := range testcases {
  var buf bytes.Buffer // Note that buf is moved into the loop. You do not need it outside the loop.
  err := tmpl.Execute(&buf, map[string]string{"Username": testcase})
  if err != nil {
    return err
  }
  filterString := buf.String()
  _, err := ldap.CompileFilter(filterString)
  if err != nil {
    return err
  }
}

// If no early return, then the filter template is considered as valid.
return nil
```

File: pkg/util/validation/formats.go [hunk 13]
```diff
@@ -209,6 +216,135 @@ func (FormatHTTPOriginSpec) CheckFormat(value interface{}) error {
...
+
+type FormatLDAPOID struct{}
+
+func (FormatLDAPOID) CheckFormat(value interface{}) error {
+	str, ok := value.(string)
+	if !ok {
+		return nil
+	}
+
+	// numericoid = number 1*( DOT number )
```
Comment (louischan-oursky) [inline pkg/util/validation/formats.go:327]: Include the relevant production rules as comment here.

```go
// numericoid = number 1*( DOT number )
// number  = DIGIT / ( LDIGIT 1*DIGIT )
// DIGIT   = %x30 / LDIGIT       ; "0"-"9"
// LDIGIT  = %x31-39             ; "1"-"9"
// https://datatracker.ietf.org/doc/html/rfc4512#section-1.4
```

Then you will see what you implement here is not strictly correct. You did not handle number with leading zeros.

Please add test cases for leading zeros.

File: pkg/lib/config/testdata/parse_secret_tests.yaml [hunk 14]
```diff
@@ -188,4 +188,24 @@ config:
...
+        type: hcaptcha # not supported
+---
+name: ldap/valid
+error: null
+config:
+  secrets:
+    - key: ldap
+      data:
+        name: test
+        username: user
```
Comment (louischan-oursky) [inline pkg/lib/config/testdata/parse_secret_tests.yaml]: I think the field is called `dn` not `username`, how come it is valid?

Also, please add more test cases, like minLength, dependentRequired. The baseline is, you test every schema you write.

File: pkg/util/validation/formats.go [hunk 15]
```diff
@@ -287,7 +285,9 @@ func (FormatLDAPSearchFilterTemplate) CheckFormat(value interface{}) error {
 		return nil
 	}
 
-	tmpl, err := template.New("search_filter").Parse(str)
+	str = strings.TrimSpace(str)
+
+	tmpl, err := template.New("search_filter").Parse(strings.Trim(str, "\n"))
```
Comment (louischan-oursky) [inline pkg/util/validation/formats.go]: I think you do not need to trim space here, you just need to trim space on the rendered result.

File: pkg/lib/config/secret.go [hunk 16]
```diff
@@ -215,6 +215,28 @@ func (c *SecretConfig) validateBotProtectionSecrets(ctx *validation.Context, bot
...
+			}
+			if !matched {
+				ctx.EmitErrorMessage(fmt.Sprintf("LDAP server user credentials for '%s' is required", server.Name))
+			} else {
+				// keys are validated by the jsonschema
+			}
+		}
+	}
+}
+
```
Comment (louischan-oursky) [inline pkg/lib/config/secret.go:239]: Add test cases for this.

**Review by louischan-oursky — CHANGES_REQUESTED** [review]

File: pkg/util/validation/formats.go [hunk 17]
```diff
@@ -325,11 +318,15 @@ func (FormatLDAPOID) CheckFormat(value interface{}) error {
...
+	// number  = DIGIT / ( LDIGIT 1*DIGIT )
+	// DIGIT   = %x30 / LDIGIT       ; "0"-"9"
+	// LDIGIT  = %x31-39             ; "1"-"9"
+	// https://datatracker.ietf.org/doc/html/rfc4512#section-1.4
 	if len(str) == 0 {
 		return errors.New("expect non-empty OID")
 	}
 
-	matched, err := regexp.MatchString(`^\d+(\.\d+)*$`, str)
+	matched, err := regexp.MatchString(`^[1-9](\.\d+)*$`, str)
```
Comment (louischan-oursky) [inline pkg/util/validation/formats.go:329]: This is still incorrect. For example, `1.01` is valid against this regex, but invalid as an oid.

File: pkg/util/validation/formats_test.go [hunk 18]
```diff
@@ -189,7 +192,9 @@ func TestFormatLDAPOID(t *testing.T) {
 		So(f(""), ShouldBeError, "expect non-empty OID")
 		So(f(".1.1"), ShouldBeError, "invalid OID")
 		So(f("1.1."), ShouldBeError, "invalid OID")
+		So(f("0.1.3"), ShouldBeError, "invalid OID")
```
Comment (louischan-oursky) [inline pkg/util/validation/formats_test.go:195]: `0.1.3` is actually a valid oid. This testcase is incorrect.

File: pkg/lib/config/testdata/secret_config_validate_tests.yaml [hunk 19]
```diff
@@ -398,8 +398,10 @@ app_config:
           search_filter_template: |
             {{if eq .Username "test@test.com"}}
                 (mail={{.Username}})
+            {{else}}
+                (uid={{.Username}})
             {{end}}
-          user_id_attribute_oid: "0.9.2342.19200300.100.1.1"
+          user_id_attribute_oid: "1.9.2342.19200300.100.1.1"
```
Comment (louischan-oursky) [inline pkg/lib/config/testdata/secret_config_validate_tests.yaml:404]: Revert this change

File: pkg/lib/config/testdata/secret_config_validate_tests.yaml [hunk 20]
```diff
@@ -431,8 +433,10 @@ app_config:
...
```
Comment (louischan-oursky) [inline pkg/lib/config/testdata/secret_config_validate_tests.yaml:439]: Revert this line

File: pkg/lib/config/testdata/secret_config_validate_tests.yaml [hunk 21]
```diff
@@ -461,7 +465,7 @@ app_config:
           url: "ldap://localhost:389"
           base_dn: "dc=localhost"
           search_filter_template: (mail={{.Username}})
-          user_id_attribute_oid: "0.9.2342.19200300.100.1.1"
+          user_id_attribute_oid: "1.9.2342.19200300.100.1.1"
```
Comment (louischan-oursky) [inline pkg/lib/config/testdata/secret_config_validate_tests.yaml:468]: Revert this line

**Review by louischan-oursky — CHANGES_REQUESTED** [review]

**Comment by yyorkchan** [comment]
Thanks @louischan-oursky

File: pkg/lib/config/testdata/parse_secret_tests.yaml [hunk 22]
```diff
@@ -188,4 +188,96 @@ config:
...
+    - key: ldap
+      data:
+        items:
+          - dn: dc=test
+            password: secret
+---
+name: ldap/missing-dn-when-password-present
+error: |-
+  invalid secrets:
+  /secrets/0/data/items/0: dependentRequired
```
Comment (louischan-oursky) [inline pkg/lib/config/testdata/parse_secret_tests.yaml:232]: When we use dependentRequired, the error shape changed a bit. This is unexpected. Let's use allOf and if then to implement the condition.


```
{
  "allOf": [
    {
      "if": {
         "required": ["dn"]
       },
       "then": {
          "required": ["password"]
       },
    },
    {
      "if": {
         "required": ["password"]
       },
       "then": {
          "required": ["dn"]
       },
    },
  ]
}

```

File: pkg/lib/config/testdata/parse_secret_tests.yaml [hunk 23]
```diff
@@ -188,4 +188,96 @@ config:
...
+    map[Required:map[password:map[actual:[name password] expected:[dn] missing:[dn]]]]
```
Comment (louischan-oursky) [inline pkg/lib/config/testdata/parse_secret_tests.yaml:233]: Note that the map has `Required`, which is not compatible with other error.

File: pkg/util/validation/formats_test.go [hunk 24]
```diff
@@ -119,6 +119,87 @@ func TestFormatHTTPOriginSpec(t *testing.T) {
...
+		So(f("1.1"), ShouldBeNil)
+		So(f("1.1.1"), ShouldBeNil)
+		So(f("0.1.3"), ShouldBeNil)
+		So(f("10.13.10"), ShouldBeNil)
+		So(f("0.9.2342.19200300.100.1.1"), ShouldBeNil)
+		So(f(""), ShouldBeError, "expect non-empty OID")
+		So(f("1"), ShouldBeError, "invalid OID")
+		So(f(".1.1"), ShouldBeError, "invalid OID")
+		So(f("1.1."), ShouldBeError, "invalid OID")
+		So(f("1.1.1.a.2"), ShouldBeError, "invalid OID")
```
Comment (louischan-oursky) [inline pkg/util/validation/formats_test.go:198]: Add a test case to show your regex can handle `0.01`

**Review by louischan-oursky — CHANGES_REQUESTED** [review]

File: pkg/lib/config/ldap.go [hunk 25]
```diff
@@ -23,21 +23,21 @@ var _ = Schema.Add("LDAPServerConfig", `
...
 	"additionalProperties": false,
-	"required": ["name", "url", "base_dn", "search_filter_template", "user_id_attribute_oid"],
+	"required": ["name", "url", "base_dn", "search_filter_template", "user_id_attribute_name"],
 	"properties": {
 		"name": { "type": "string", "minLength": 1 },
 		"url": { "type": "string", "format": "ldap_url" },
 		"base_dn": { "type": "string", "format": "ldap_dn" },
 		"search_filter_template": { "type": "string", "format": "ldap_search_filter_template" },
-		"user_id_attribute_oid": { "type": "string", "format": "ldap_oid" }
+		"user_id_attribute_name": { "type": "string", "format": "ldap_attribute" }
```
Comment (louischan-oursky) [inline pkg/lib/config/ldap.go:32]: The format is called `ldap_attribute_name`

File: pkg/util/validation/formats.go [hunk 26]
```diff
@@ -309,29 +309,29 @@ func (FormatLDAPSearchFilterTemplate) CheckFormat(value interface{}) error {
...
+		return errors.New("expect non-empty attribute")
 	}
 
-	matched, err := regexp.MatchString(`^(?:[0-9]|[1-9][0-9]+)(?:\.(?:[0-9]|[1-9][0-9]+))+$`, str)
+	// An attribute description option is represented by the ABNF:
+	// options = *( SEMI option )
+	// option = 1*keychar
+	// keychar = ALPHA / DIGIT / HYPHEN
+	// According to https://datatracker.ietf.org/doc/html/rfc4512#section-2.5
+	matched, err := regexp.MatchString(`^[a-zA-Z\d-]+$`, str)
```
Comment (louischan-oursky) [inline pkg/util/validation/formats.go:329]: But attribute name starting with a hyphen is uncommon. Let's use production rule of `keystring` here. Update the comment to reflect this implementation choice.

File: pkg/util/validation/formats_test.go [hunk 27]
```diff
@@ -182,20 +182,22 @@ func TestFormatLDAPSearchFilterTemplate(t *testing.T) {
...
+		So(f("ou"), ShouldBeNil)
+		So(f("dc"), ShouldBeNil)
+		So(f("hyphen-"), ShouldBeNil)
+		So(f("123"), ShouldBeNil)
+		So(f("abcdEFG123-"), ShouldBeNil)
+		So(f(""), ShouldBeError, "expect non-empty attribute")
+		So(f("should have no space"), ShouldBeError, "invalid attribute")
+		So(f("abcd!@#$%^&*()"), ShouldBeError, "invalid attribute")
+		So(f("dc="), ShouldBeError, "invalid attribute")
+		So(f("dc=example,dc=com"), ShouldBeError, "invalid attribute")
```
Comment (louischan-oursky) [inline pkg/util/validation/formats_test.go:200]: Add a test to show we forbid attribute name starting with a digit or a hyphen.

**Review by louischan-oursky — CHANGES_REQUESTED** [review]

**Review by louischan-oursky — APPROVED** [review]

**Pull request merged**